File: python/binary_tree_right_side_view.py
from collections import deque
from typing import Deque, Optional
import sys

sys.path.append("./models")
from binary_tree_node import TreeNode

sys.path.append("./utils")
from deserialize_to_binary_tree import DeserializeFromList


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> list[int]:
        result: list[int] = []
        queue: Deque[Optional[TreeNode]] = deque([root])

        while len(queue) > 0:
            rightNode: Optional[TreeNode] = None
            queLen: int = len(queue)
            for _ in range(0, queLen):
                ele: Optional[TreeNode] = queue.popleft()
                if ele != None:
                    rightNode = ele
                    queue.append(ele.left)
                    queue.append(ele.right)

            if rightNode != None:
                result.append(rightNode.val)

        return result


def test_rightSideView_case_1():
    # arrange
    arr: list[Optional[int]] = [1, 2, 3, None, 5, None, 4]
    root: Optional[TreeNode] = DeserializeFromList(arr)
    expected: list[int] = [1, 3, 4]

    # act
    solution = Solution()
    actual = solution.rightSideView(root)

    # assert
    assert actual == expected


def test_rightSideView_case_2():
    # arrange
    arr: list[Optional[int]] = [1, None, 3]
    root: Optional[TreeNode] = DeserializeFromList(arr)
    expected: list[int] = [1, 3]

    # act
    solution = Solution()
    actual = solution.rightSideView(root)

    # assert
    assert actual == expected


def test_rightSideView_case_3():
    # arrange
    arr: list[Optional[int]] = []
    root: Optional[TreeNode] = DeserializeFromList(arr)
    expected: list[int] = []

    # act
    solution = Solution()
    actual = solution.rightSideView(root)

    # assert
    assert actual == expected

File: rust/src/code/binary_tree_right_side_view.rs
use crate::code::models::binary_tree_node::TreeNode;
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

struct Solution {}
impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result: Vec<i32> = Vec::new();
        let mut queue: VecDeque<Option<Rc<RefCell<TreeNode>>>> = VecDeque::from([root]);

        while !queue.is_empty() {
            let curr_queue_len: usize = queue.len();
            let mut right_node: Option<Rc<RefCell<TreeNode>>> = None;
            for _ in 0..curr_queue_len {
                if let Some(node) = queue.pop_front() {
                    if let Some(e) = node.clone() {
                        right_node = node;
                        queue.push_back(e.borrow().left.clone());
                        queue.push_back(e.borrow().right.clone());
                    }
                }
            }

            if let Some(r) = right_node {
                result.push(r.borrow().val);
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;
    use crate::code::utils::deserialize_to_binary_tree::deserialize_to_BT;

    #[test]
    fn test_right_side_view_case_1() {
        // arrange
        let root = deserialize_to_BT(vec![
            Some(1),
            Some(2),
            Some(3),
            None,
            Some(5),
            None,
            Some(4),
        ]);
        let expected = vec![1, 3, 4];

        // act
        let actual = Solution::right_side_view(root);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_right_side_view_case_2() {
        // arrange
        let root = deserialize_to_BT(vec![Some(1), None, Some(3)]);
        let expected = vec![1, 3];

        // act
        let actual = Solution::right_side_view(root);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_right_side_view_case_3() {
        // arrange
        let root = deserialize_to_BT(vec![]);
        let expected = vec![];

        // act
        let actual = Solution::right_side_view(root);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_right_side_view_case_4() {
        // arrange
        let root = deserialize_to_BT(vec![Some(1), Some(2)]);
        let expected = vec![1, 2];

        // act
        let actual = Solution::right_side_view(root);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_right_side_view_case_5() {
        // arrange
        let root = deserialize_to_BT(vec![
            Some(0),
            Some(1),
            Some(2),
            None,
            Some(3),
            Some(4),
            None,
            None,
            Some(5),
            Some(9),
            None,
            None,
            Some(6),
            Some(10),
            None,
        ]);
        let expected = vec![0, 2, 4, 9, 10];

        // act
        let actual = Solution::right_side_view(root);

        // assert
        assert_eq!(actual, expected);
    }
}
